File: Regression/Polynomial_Regression/polynomial_regression.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Polynomial Regression Implementation
Updated for better portability
"""

# Importing necessary libraries
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import os
from sklearn.preprocessing import PolynomialFeatures
from sklearn.linear_model import LinearRegression

# Get the directory where the script is located
script_dir = os.path.dirname(os.path.abspath(__file__))
file_path = os.path.join(script_dir, "Position_Salaries.csv")  # Dataset file

# Checking if the dataset exists
if os.path.isfile(file_path):
    dataset = pd.read_csv(file_path)
    print("Data imported successfully!")  
else:
    raise FileNotFoundError(f"Error: File not found: {file_path}. Please make sure 'Position_Salaries.csv' is in the same directory.")

# Splitting dataset into independent (X) and dependent (y) variables
X = dataset.iloc[:, 1:-1].values  # Position Level (independent variable)
y = dataset.iloc[:, -1].values   # Salary (dependent variable)

# Fitting Linear Regression to the dataset for comparison
lin_reg = LinearRegression()
lin_reg.fit(X, y)

# Fitting Polynomial Regression to the dataset
poly_reg = PolynomialFeatures(degree=4)  # Degree 4 for better curve fitting
X_poly = poly_reg.fit_transform(X)
lin_reg2 = LinearRegression()
lin_reg2.fit(X_poly, y)

# Visualizing the Linear Regression results
plt.scatter(X, y, color='red')
plt.plot(X, lin_reg.predict(X), color='blue')
plt.title('Truth or Bluff (Linear Regression)')
plt.xlabel('Position Level')
plt.ylabel('Salary')
plt.show()

# Visualizing the Polynomial Regression results
plt.scatter(X, y, color='red')
plt.plot(X, lin_reg2.predict(X_poly), color='blue')
plt.title('Truth or Bluff (Polynomial Regression)')
plt.xlabel('Position Level')
plt.ylabel('Salary')
plt.show()

# Predicting a new result with Linear Regression
salary_pred_linear = lin_reg.predict([[6.5]])
print(f"Predicted Salary using Linear Regression: {salary_pred_linear[0]:.2f}")

# Predicting a new result with Polynomial Regression
salary_pred_poly = lin_reg2.predict(poly_reg.fit_transform([[6.5]]))
print(f"Predicted Salary using Polynomial Regression: {salary_pred_poly[0]:.2f}")

File: Regression/Polynomial_Regression/polynomial_regression.R
# Polynomial Regression

# Importing Dataset
library(readr)
library(ggplot2)
dataset <- read_csv('Position_Salaries.csv') 
dataset = dataset[, 2:3]  # Keep only Level and Salary columns

# Fitting Simple Linear Regression
lin_reg = lm(formula = Salary ~ Level, data = dataset)

# Fitting Polynomial Regression using poly()
poly_reg = lm(formula = Salary ~ poly(Level, 4, raw = TRUE), data = dataset)

# Model Summaries
summary(lin_reg)
summary(poly_reg)

# Visualising Linear Regression Results
ggplot() +
  geom_point(aes(x = dataset$Level, y = dataset$Salary), colour = 'red') +
  geom_line(aes(x = dataset$Level, y = predict(lin_reg, newdata = dataset)), colour = 'blue') +
  ggtitle('Truth/Bluff (Linear Regression)') +
  xlab('Level') +
  ylab('Salary')

# Visualising Polynomial Regression Results (Higher Resolution)
x_grid = seq(min(dataset$Level), max(dataset$Level), 0.1)
ggplot() +
  geom_point(aes(x = dataset$Level, y = dataset$Salary), colour = 'red') +
  geom_line(aes(x = x_grid, y = predict(poly_reg, newdata = data.frame(Level = x_grid))), colour = 'blue') +
  ggtitle('Truth/Bluff (Polynomial Regression)') +
  xlab('Level') +
  ylab('Salary')

# Predicting a new result with Linear Regression
y_pred = predict(lin_reg, data.frame(Level = 6.5))

# Predicting a new result with Polynomial Regression
y_pred2 = predict(poly_reg, data.frame(Level = 6.5))

# Model Evaluation (Mean Squared Error)
mse_lin = mean((dataset$Salary - predict(lin_reg, dataset))^2)
mse_poly = mean((dataset$Salary - predict(poly_reg, dataset))^2)

print(paste("Linear Regression MSE:", round(mse_lin, 2)))
print(paste("Polynomial Regression MSE:", round(mse_poly, 2)))
